website/blog/2024-02-08-kubernetes-frontends/index.mdx

```diff
@@ -0,0 +1,195 @@
+---
+slug: kubernetes-frontends
+title: The inner workings of kubernetes management frontends - A software engineers' perspective
+authors: [ christophenne ]
+tags: [ glasskube, kubernetes, frontends ]
+---
+
+# The inner workings of kubernetes management frontends - A software engineers' perspective
+
+The rise of Kubernetes in recent years has led to an astonishing number of open-source Kubernetes management tools seemingly appearing out of nowhere.
+The goal of the research leading to this article was to merely understand the architecture of some of these tools and to subsequently provide a brief overview
+and options for developers trying to get started with their own Kubernetes frontend.
+We will not dive deep into the actual tools and what problems they are trying to solve, but instead focus on the software engineering aspect.
+We are also exclusively exploring open-source and self-hosted tools and leaving the PaaS/IaaS platforms of cloud providers aside - that would be a whole different article.
+
+Setting up and interacting with your first cluster can be overwhelming. Just like me, you might have come across the infamous [`kubernetes/dashboard`](https://github.com/kubernetes/dashboard),
+followed the installation instructions, and asked yourself: "What did I just do and why exactly does this work the way it works?"
+And after some tinkering with your cluster, you might have installed even more external tools that help you with some specific aspects of cluster management,
+providing you with either a CLI or a Web UI.
+
+As a software engineer mostly engaged in web development in recent years, I was curious about how these kinds of tools are built and deployed.
+
+Let's first clarify some of the basics needed for the following exploration of different Kubernetes UIs. After that, we will see what they have in common and what makes this kind of software special, to finally form a recommendation of how one could build a Kubernetes Web UI themselves.
+
+## Kubernetes Basics
+
+The [official documentation](https://kubernetes.io/docs/concepts/overview/components/) is more than helpful anyway; there is just one important thing to remember:
+Whenever and wherever you interact with your cluster, you do it via the Kubernetes API - this holds true at least for the scope of this article,
+though there might be other use cases.
+
+As a consumer of this API, one needs to know where it is hosted and how to authenticate against it.
+The Kubernetes API can be accessed both from inside a cluster (i.e., from an application running on a pod) and outside a cluster (e.g., from your command line).
```

Comment:
This is often true, but not always. In many cases, the cluster API is only available from within a VPN or behind a strict firewall (e.g. Amazon EKS). Maybe this could be mentioned here briefly.